README.md
<div id="top"></div>

# Judo Spin

360° degrees image spinner and rotation built with vanilla JavaScript. 

## [View Online Demo](https://rodgath.github.io/judo-spin/demo/)

<br>

[![NPM Version](https://img.shields.io/npm/v/judo-spin.svg?color=00BAB8)]()
[![NPM License](https://img.shields.io/npm/l/all-contributors.svg?color=00a7a6)](https://github.com/Rodgath/judo-spin/blob/master/LICENSE)
[![NPM Downloads](https://img.shields.io/npm/dw/judo-spin.svg?color=009593)]() 
![npm bundle size (version)](https://img.shields.io/bundlephobia/minzip/judo-spin/latest?color=008281)

![GitHub issues](https://img.shields.io/github/issues-raw/Rodgath/judo-spin?color=37bf4c)
[![GitHub pull requests](https://img.shields.io/github/issues-pr/Rodgath/judo-spin.svg?color=32ac44)]()
[![PR's Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?color=278635)](http://makeapullrequest.com) 

# Table of Contents

- [Installation](#installation)
- [Implementation](#implementation)
  - [Method 1](#method-1)
  - [Method 2](#method-2)
  - [Method 3](#method-3)
- [Options](#options)
- [License](#license)

# Installation 

**[A]** Download 

+ [judo-spin.min.js](https://unpkg.com/judo-spin@latest/dist/js/judo-spin.min.js) - Minified
+ [judo-spin.js](https://unpkg.com/judo-spin@latest/dist/js/judo-spin.min.js) - Unminified

**[B]** Package

* Install with [yarn](https://yarnpkg.com/): `yarn add judo-spin`

* Install with [npm](https://www.npmjs.com/package/judo-spin): `npm install judo-spin` 

* [![NPM](https://nodei.co/npm/judo-spin.png?downloads=true)](https://www.npmjs.com/package/judo-spin) 

**[C]** Get a local working copy of the development repository _(Optional)_ <br />
`git clone https://github.com/Rodgath/judo-spin.git`


# Implementation
There are several methods you can use to add Judo Spin into your project.

## Method 1

#### Using `data-judo-spin` attribute. [View Demo](https://rodgath.github.io/judo-spin/demo/index2.html)
Your HTML code of upto 36 images
```html
<div data-judo-spin='{ "currImage": "6" }'>
  <img src="images/car/01.png" alt="Image 01">
  <img src="images/car/02.png" alt="Image 02">

  <!-- More Images Here -->
  
  <img src="images/car/35.png" alt="Image 35">
  <img src="images/car/36.png" alt="Image 36">
</div>
```

#### Enqueue the judoSpin script at the bottom of your markup
+ Using local script file.
```html
<script src="judo-spin.min.js"></script>
<!-- OR -->
<script src="./node_modules/judo-spin/dist/js/judo-spin.min.js"></script>
```
+ Using CDN file. _(Optional)_
```html
<script src="https://cdn.jsdelivr.net/npm/judo-spin@latest/dist/js/judo-spin.min.js"></script>
```

##### <div align="right"><a href="#top">&uarr; TOP &uarr;</a></div>

---
## Method 2

#### Initializing with `judoSpin` function. [View Demo](https://rodgath.github.io/judo-spin/demo/index.html)
Your HTML code of upto 36 images
```html
<div class="image-spin-box">
  <img src="images/car/01.png" alt="Image 01">
  <img src="images/car/02.png" alt="Image 02">

  <!-- More Images Here -->
  
  <img src="images/car/35.png" alt="Image 35">
  <img src="images/car/36.png" alt="Image 36">
</div>
```

#### Call the `judoSpin()` function with two arguments.
...**1)** The element _'class'_ or _'id'_ holding the images 

...**2)** The _'options'_ object
```javascript
document.addEventListener('DOMContentLoaded', function() {
  judoSpin('.image-spin-box', { currImage: 1, enableDragHangle: true });
});
```

#### Enqueue the judoSpin script at the bottom of your markup
+ Using local script file.
```html
<script src="judo-spin.min.js"></script>
<!-- OR -->
<script src="./node_modules/judo-spin/dist/js/judo-spin.min.js"></script>
```
+ Using CDN file. _(Optional)_
```html
<script src="https://cdn.jsdelivr.net/npm/judo-spin@latest/dist/js/judo-spin.min.js"></script>
```

##### <div align="right"><a href="#top">&uarr; TOP &uarr;</a></div>

---
## Method 3

#### Using JSON object of images
...**[A]** - Adding the JSON object inside the `data-judo-spin` attribute. [View Demo](https://rodgath.github.io/judo-spin/demo/index3.html)

```html
<div data-judo-spin='{ 
    "currImage": 13,
    "images": [
      { "src": "images/car/01.png", "title": "Image 01" },
      { "src": "images/car/02.png", "title": "Image 02" },
      
      // More Images Here

      { "src": "images/car/35.png", "title": "Image 35" },
      { "src": "images/car/36.png", "title": "Image 36" }
    ] 
  }'></div>
```

...**[B]** - Add the JSON object inside the `judoSpin` function. [View Demo](https://rodgath.github.io/judo-spin/demo/index4.html)

HTML code.
```html
<div class="image-rotation-box image-spin-demo"></div>
```
JavaScript code

```javascript
document.addEventListener('DOMContentLoaded', function() {
   judoSpin('.image-rotation-box', {
      currImage: 22, 
      images: [
        { "src": "images/car/01.png", "title": "Image 01" },
        { "src": "images/car/02.png", "title": "Image 02" },
        
        // More Images Here

        { "src": "images/car/35.png", "title": "Image 35" },
        { "src": "images/car/36.png", "title": "Image 36" }
      ]
    });
});
```
##### <div align="right"><a href="#top">&uarr; TOP &uarr;</a></div>

---
# Options 

Name | Type | Default | Description
:--- | :--- | :------ | :----------
`currImage` | _number_ | 1 | The default image when judo spin loads. Range is _1 - 36_.
`images` | _array_|_object_ | - | Array of images as JSON object or DOM images.
`enableDragHandle` | _boolean_ | false | Whether to show the drag handler scroller.

## License
judoSpin is an open-source project released under the [MIT license](http://www.opensource.org/licenses/mit-license.php).

---

Crafted by [Rodgath](https://twitter.com/Rodgath)
##### <div align="right"><a href="#top">&uarr; TOP &uarr;</a></div>

CHANGELOG.md
###### v1.0.3 - Aug 15, 2023
```
= Update README
```
###### v1.0.2 - Aug 8, 2023
```
= Fix: Deduct 1 from 'initialImageIndex'
```
###### v1.0.1 - Aug 8, 2023
```
= Update draggable button layout 
= Add 'enableDragHandle' options argument
= Set initial draggable button position
= Untrack .vscode/settings.json file
```
###### v1.0.0 - Aug 5, 2023
```
= Initial release.
```

gulpfile.js
const { src, dest, series, parallel, watch, task } = require('gulp');
const uglify = require('gulp-uglify');
const babel = require('gulp-babel');
const rename = require('gulp-rename');
const header = require('gulp-header');
const shell = require('gulp-shell');
const pkg = require('./package.json');

const sourceBranch = 'main';
const targetBranch = 'gh-pages';
const folderToCopy = 'demo';

/* Copy folder between branches */
const copyFolder = () => {
  return shell.task([
    `git checkout ${targetBranch}`,
    `git checkout ${sourceBranch} -- ${folderToCopy}`,
    `git add ${folderToCopy}`,
    `git commit -m "Copy '${folderToCopy}' folder from ${sourceBranch}"`,
    `git push origin ${targetBranch}`,
    `git checkout ${sourceBranch}`
  ]);
}

/* Set banner for dist files */
const setBanner = () => {
	const banner = [
    '/**',
    ` * Judo Spin`,
    ` * @name        ${pkg.name}`,
    ` * @description ${pkg.description}`,
    ` * @link        ${pkg.homepage}`,
    ` * @author      ${pkg.author.name}, ${pkg.author.web}`,
    ` * @version     v${pkg.version}`,
    ` * @created     Jul 20, 2023`,
    ` * @updated     Aug 15, ${new Date().getFullYear()}`,
    ` * @copyright   Copyright (C) 2023-${new Date().getFullYear()}, ${pkg.author.name}`,
    ` * @license     ${pkg.license}`,
    ` * @licenseMIT  ${pkg.homepage}/blob/main/LICENSE`,
    ` * @demoExample https://rodgath.github.io/judo-spin/demo/`,  
    ' */',
    ''
  ].join('\n');
  
	return src('./dist/{css,js}/**/*.{js,css}')
  .pipe(header(banner))
  .pipe(dest('./dist'))
}

const compressJs = () => {
  return src('src/*.js')
        .pipe(dest('./demo/js')) // Send original script file to /demo
        .pipe(dest('./dist/js')) // Send original script file to /dist
        .pipe(babel({
            presets: ['@babel/env']
        }))
        .pipe(uglify())
        .pipe(rename( { suffix: '.min' } ))
        .pipe(dest('./demo/js'))
        .pipe(dest('./dist/js'))
}

const watchTask = () => {
  watch(['./src/*.js', './demo/*.html'], { events: 'all' }, series(compressJs, setBanner))
}

const buildTask = () => {
  return series(compressJs, setBanner)
}

/* Tasks */
task('copy:demo', copyFolder());
task('set:banner', setBanner);
task('compress:js', compressJs);
task('watch', watchTask);
task('build', buildTask());

exports.default = buildTask();

demo/js/script.js
/*  Get all the accordion headers */
const accordionHeaders = document.querySelectorAll('.accordion-header');

/*  Loop through each header and add a click event listener */
accordionHeaders.forEach(header => {
  header.addEventListener('click', function () {

    /* Get the corresponding content for this header */
    const content = this.nextElementSibling;

    /* Toggle the accordion content with slide effect */
    toggleAccordionContent(header, content);
  });
});

/* Toggle the accordion content with slide effect */
function toggleAccordionContent(header, content) {
  let heightAdjust = 8; // Adjust the speed of the slide by changing this value
  let startHeight = 0;
  const maxHeight = content.scrollHeight;
  
  function expandAccordion() {

    header.classList.add('active');

    content.classList.remove('hidden');

    startHeight += heightAdjust;

    if (startHeight < maxHeight) {
      content.style.height = `${startHeight}px`;
      requestAnimationFrame(expandAccordion);
    } else {
      content.style.height = null;
    }
  }

  function collapseAccordion() {
    startHeight -= heightAdjust;

    if (Math.abs(startHeight) < maxHeight) {
      content.style.height = maxHeight - Math.abs(startHeight) + 'px';
      requestAnimationFrame(collapseAccordion);
    } else {
      content.style.height = null;
      header.classList.remove('active');
      content.classList.add('hidden');
    }
  }

  if (content.classList.contains('hidden')) {
    expandAccordion();
  } else {
    collapseAccordion();
  }
}

src/image-spin.js

function imageSpin(element, options) {

  var defaults = {
    currImg: 1,
  };

  options = Object.assign({}, defaults, options);
    
  /* Current image starts at 0 or based on options entry */
  let currImagePos = convertCurrImageNumber(options.currImage) - 1;

  /* Get the image box element */
  const imageSpinBox = document.querySelector(element);

  /* Get all the images inside the spinbox element */
  const imagesToSpin = imageSpinBox.querySelectorAll('img');
  const totalImages = imagesToSpin.length;

  /* Calculate the angle between each image */
  const anglePerImage = 360 / totalImages;

  /* Set the initial angle to 0 or based on current mage*/
  let currentAngle = currImagePos > 0 ? currImagePos * anglePerImage : 0;

  /* Flag to track if the motion should start */
  let motionStarted = false;

  /* Create the overlay div element */
  const overlayElement = document.createElement('div');
  
  /* Convert a string to a positive number or its absolute value if negative */
  function convertCurrImageNumber(str) {
    const number = parseInt(str, 10); // Convert string to integer

    if (!isNaN(number)) {
      return Math.abs(number);
    } else {
      return 1;
    }
  }

  /* Set the attributes and styles for the overlay element */
  (function(overlayElement) {
    overlayElement.className = 'image-spin-overlay';
    overlayElement.style.position = 'absolute';
    overlayElement.style.width = '100%';
    overlayElement.style.height = '100%';
    overlayElement.style.top = '0';
    overlayElement.style.left = '0';
    overlayElement.style.zIndex = '1';

    imageSpinBox.appendChild(overlayElement);

    /* Add event listeners to change cursor on hover and reset on mouseout */
    overlayElement.addEventListener('mouseover', () => overlayElement.style.cursor = 'e-resize');
    overlayElement.addEventListener('mouseout', () => overlayElement.style.cursor = 'auto');

  })(overlayElement);

  /* Set image spin box attributes and styles */
  (function(imageSpinBox) {
    imageSpinBox.classList.add('image-spin-box-container');
    imageSpinBox.style.position = 'relative';
    imageSpinBox.style.width = '100%';
    imageSpinBox.style.height = 'auto';
    imageSpinBox.style.overflow = 'hidden';
  })(imageSpinBox);
  
  /* Set images' attributes and styles */
  (function(imageSpinBox) {
    const images = imageSpinBox.getElementsByTagName('img');

    for (let i = 0; i < images.length; i++) {
      const image = images[i];
      image.style.width = '100%';
    }
  })(imageSpinBox);

  /* Handle mousedown event */
  const handleMouseDown = event => invokeMotion(event);

  /* Handle touchstart event */
  const handleTouchStart = event => invokeMotion(event);

  /* Listen for mousedown and touchstart events on the overlayElement element */
  overlayElement.addEventListener('mousedown', handleMouseDown);
  overlayElement.addEventListener('touchstart', handleTouchStart);

  /* Invoke image spin motion based on mousemove and touchmove events */
  const invokeMotion = event => {

    event.preventDefault();

    const _event = event.type === 'touchmove' ? 'touchmove' : 'mousemove';
    const _listener = event.type === 'touchmove' ? handleTouchMove : handleMouseMove;
    
    /* Start motion */
    motionStarted = true;

    /* Get the index of the current visible image */
    const currentIndex = calculateImageIndex(currentAngle);
    
    /* Calculate the initial angle based on the current visible image */
    currentAngle = currentIndex * anglePerImage;

    /* Listen for touchmove event */
    overlayElement.addEventListener(_event, _listener);
  }

  /* Handle mouseup event */
  const handleMouseUp = () => {
    motionStarted = false;
    overlayElement.removeEventListener('mousemove', handleMouseMove);
  }

  /* Handle touchend event */
  const handleTouchEnd = () => {
    motionStarted = false;
    overlayElement.removeEventListener('touchmove', handleTouchMove);
  }

  /* Add event listeners for mouseup and touchend events */
  document.addEventListener('mouseup', handleMouseUp);
  document.addEventListener('touchend', handleTouchEnd);

  /* Handle mousemove event */
  const handleMouseMove = event => handleMovement(event);

  /* Handle touchmove event */
  const handleTouchMove = event => handleMovement(event);

  /* Handle image spin movement based on event */
  const handleMovement = event => {
    if (motionStarted) {
      const _clientX = event.type === 'touchmove' ? event.touches[0].clientX : event.clientX;
      const _clientY = event.type === 'touchmove' ? event.touches[0].clientY : event.clientY;

      const dir = detectDirection(_clientX, _clientY);

      dir === 'right' ? currImagePos++ : dir === 'left' ? currImagePos-- : null;

      /* Calculate the rotation angle based on mouse position */
      const maxRotation = 360; // 360 degrees for a full rotation; Maybe alterable
      currentAngle = currImagePos * anglePerImage;
      
      /* Adjust currentAngle to be within the range of 0 to 359 degrees, without negavtive numbers */
      currentAngle = (currentAngle % maxRotation + maxRotation) % maxRotation;

      showImageForAngle(currentAngle);
    }
  }
  
  /* Calculate the current image index based on the current angle */
  const calculateImageIndex = angle => {

    let index = Math.floor(angle / anglePerImage);

    if (index < 0) {
        index = totalImages - 1;
    } else if (index >= totalImages) {
        index = 0;
    }
    return index;
  }

  /* Show the image based on the current angle */
  const showImageForAngle = angle => {

      /* get the index of the image to show based on the angle and total images */
      const imageIndex = Math.floor(angle / anglePerImage) + 1;

      /* Show the corresponding image and hide the others */
      for (let i = 1; i <= totalImages; i++) {
        const img = imageSpinBox.querySelector('img:nth-child(' + i + ')');
        img.style.display = (i === imageIndex) ? 'block' : 'none';
      }
  }
  
  /* Show the preferred/intended image by default */
  showImageForAngle(currentAngle);

  /* Store the previous position of the mouse */
  let prevX = null;
  let prevY = null;

  /* Mouse sensitivity factor (adjust this to set sensitivity) */
  const sensitivity = 1; // Increase for higher sensitivity, decrease for lower sensitivity

  /* Detect the direction */
  function detectDirection(currentX, currentY) {
    let direction;
    if (prevX !== null && prevY !== null) {
      const deltaX = currentX - prevX;

      if (deltaX > sensitivity) {
        direction = 'right';
      } else if (deltaX < -sensitivity) {
        direction = 'left';
      }
    }

    /* Update the previous position to the current position */
    prevX = currentX;
    prevY = currentY;

    return direction;
  }
}

demo/js/judo-spin.js
function judoSpin(element, options) {

  var defaults = {
    currImage: 1,
    images: [],
    enableDragHandle: false
  };
  
  /* Check if options argument is provided directly or in the data attribute */
  if (!options) {
    var dataAttributeOptions = element.dataset.judoSpin;
    if (dataAttributeOptions) {
      options = JSON.parse(dataAttributeOptions);
    }
  }

  /* Merge 'options' with 'defaults' */
  options = Object.assign({}, defaults, options);
    
  /* Current image starts at 0 or based on options entry */
  let currImagePos = convertCurrImageNumber(options.currImage) - 1;
    
  /* Flag to track if the button is being dragged */
  let dragging = false;

  /* Set judo box element */
  let judoSpinBox;

  /* Set judo box wrapper element */
  let judoSpinWrapper;
  
  /* Check if 'element' is a valid DOM element */
  if (element instanceof HTMLElement || element instanceof Node) {
    judoSpinBox = element;
  } else {
    judoSpinBox = document.querySelector(element);
  }
  
  /* Get all the images inside the spinbox element */
  const imagesToSpin = getImagesToSpin();
  const totalImages = imagesToSpin.length;

  /* Calculate the angle between each image */
  const anglePerImage = 360 / totalImages;

  /* Set the initial angle to 0 or based on current mage*/
  let currentAngle = currImagePos > 0 ? currImagePos * anglePerImage : 0;

  /* Flag to track if the motion should start */
  let motionStarted = false;

  /* Create the overlay div element */
  const overlayElement = document.createElement('div');
  
  /* Create Judo scroller element */
  const judoScroller = options.enableDragHandle ? document.createElement('div') : null;
  
  /* Create Judo draggable element */
  const judoDraggable = options.enableDragHandle ? document.createElement('div') : null;
  
  /* Convert a string to a positive number or its absolute value if negative */
  function convertCurrImageNumber(str) {
    const number = parseInt(str, 10); // Convert string to integer

    if (!isNaN(number)) {
      return Math.abs(number);
    } else {
      return 1;
    }
  }

  /* Get the images to spin */
  function getImagesToSpin() {

    let spinImages;
    
    /* Check if ihe "images" property exists, is an array, and has items */
    if (options.hasOwnProperty('images') && Array.isArray(options.images) && options.images.length > 0) {
      
      options.images.forEach(imageObj => {
        const imgElement = document.createElement('img');
        imgElement.src = imageObj.src;
        imgElement.alt = imageObj.title;
        
        judoSpinBox.appendChild(imgElement);
      });

      spinImages = judoSpinBox.querySelectorAll('img')
    } else {
      spinImages = judoSpinBox.querySelectorAll('img')
    }

    return spinImages;
  }

  /* Set the attributes and styles for the overlay element */
  (function(overlayElement) {
    overlayElement.className = 'judo-spin-overlay';
    overlayElement.style.position = 'absolute';
    overlayElement.style.width = '100%';
    overlayElement.style.height = '100%';
    overlayElement.style.top = '0';
    overlayElement.style.left = '0';
    overlayElement.style.zIndex = '1';

    judoSpinBox.appendChild(overlayElement);

    /* Add event listeners to change cursor on hover and reset on mouseout */
    overlayElement.addEventListener('mouseover', () => overlayElement.style.cursor = 'e-resize');
    overlayElement.addEventListener('mouseout', () => overlayElement.style.cursor = 'auto');

  })(overlayElement);

  /* Set Judo spin box attributes and styles */
  (function(judoSpinBox) {
    judoSpinBox.classList.add('judo-spin-container');
    judoSpinBox.style.position = 'relative';
    judoSpinBox.style.width = '100%';
    judoSpinBox.style.height = 'auto';
    judoSpinBox.style.overflow = 'hidden';
  })(judoSpinBox);
  
  /* Set images' attributes and styles */
  (function(judoSpinBox) {
    const images = judoSpinBox.getElementsByTagName('img');

    for (let i = 0; i < images.length; i++) {
      const image = images[i];
      image.style.width = '100%';
    }
  })(judoSpinBox);

  /* Create judoSpinBox wrapper */
  (function(judoSpinBox) {
    judoSpinWrapper = document.createElement('div');
    judoSpinWrapper.classList.add('judo-spin-wrapper');
    judoSpinWrapper.style.width = 'fit-content';
    judoSpinWrapper.style.margin = '0 auto';
    judoSpinBox.parentNode.insertBefore(judoSpinWrapper, judoSpinBox);
    judoSpinWrapper.appendChild(judoSpinBox);
  })(judoSpinBox);
  
  /* Handle mousedown event */
  const handleMouseDown = event => invokeMotion(event);

  /* Handle touchstart event */
  const handleTouchStart = event => invokeMotion(event);

  /* Listen for mousedown and touchstart events on the overlayElement element */
  overlayElement.addEventListener('mousedown', handleMouseDown);
  overlayElement.addEventListener('touchstart', handleTouchStart);

  /* Invoke image spin motion based on mousemove and touchmove events */
  const invokeMotion = event => {

    event.preventDefault();

    const _event = event.type === 'touchstart' ? 'touchmove' : 'mousemove';
    const _listener = event.type === 'touchstart' ? handleTouchMove : handleMouseMove;

    console.log(event.type);
    
    /* Start motion */
    motionStarted = true;

    /* Get the index of the current visible image */
    const currentIndex = calculateImageIndex(currentAngle);
    
    /* Calculate the initial angle based on the current visible image */
    currentAngle = currentIndex * anglePerImage;

    /* Listen for touchmove event */
    overlayElement.addEventListener(_event, _listener);
  }

  /* Handle mouseup event */
  const handleMouseUp = () => {
    motionStarted = false;
    overlayElement.removeEventListener('mousemove', handleMouseMove);
  }

  /* Handle touchend event */
  const handleTouchEnd = () => {
    motionStarted = false;
    overlayElement.removeEventListener('touchmove', handleTouchMove);
  }

  /* Add event listeners for mouseup and touchend events */
  document.addEventListener('mouseup', handleMouseUp);
  document.addEventListener('touchend', handleTouchEnd);

  /* Handle mousemove event */
  const handleMouseMove = event => handleMovement(event);

  /* Handle touchmove event */
  const handleTouchMove = event => handleMovement(event);

  /* Handle image spin movement based on event */
  const handleMovement = event => {
    if (motionStarted) {
      const _clientX = event.type === 'touchmove' ? event.touches[0].clientX : event.clientX;
      const _clientY = event.type === 'touchmove' ? event.touches[0].clientY : event.clientY;

      const dir = detectDirection(_clientX, _clientY);

      dir === 'right' ? currImagePos++ : dir === 'left' ? currImagePos-- : null;

      /* Calculate the rotation angle based on mouse or touch position */
      const maxRotation = 360; // 360 degrees for a full rotation; Maybe alterable
      currentAngle = currImagePos * anglePerImage;
      
      /* Adjust currentAngle to be within the range of 0 to 359 degrees, without negavtive numbers */
      currentAngle = (currentAngle % maxRotation + maxRotation) % maxRotation;

      showImageForAngle(currentAngle);
        
      /* Call handleDragMove to update the draggable button position */
      if (options.enableDragHandle) {
        dragging = true;
        handleDragMove(event);
      }
    }
  }
  
  /* Calculate the current image index based on the current angle */
  const calculateImageIndex = angle => {

    let index = Math.floor(angle / anglePerImage);

    if (index < 0) {
        index = totalImages - 1;
    } else if (index >= totalImages) {
        index = 0;
    }
    return index;
  }

  /* Show the image based on the current angle */
  const showImageForAngle = angle => {

      /* get the index of the image to show based on the angle and total images */
      const imageIndex = Math.floor(angle / anglePerImage) + 1;

      /* Show the corresponding image and hide the others */
      for (let i = 1; i <= totalImages; i++) {
        const img = judoSpinBox.querySelector('img:nth-child(' + i + ')');
        img.style.display = (i === imageIndex) ? 'block' : 'none';
      }
  }
  
  /* Show the preferred/intended image by default */
  showImageForAngle(currentAngle);

  /* Store the previous position of the mouse/touch */
  let prevX = null;
  let prevY = null;

  /* Mouse/touch sensitivity factor */
  const sensitivity = 1; // Increase for higher sensitivity, decrease for lower sensitivity

  /* Detect the direction */
  function detectDirection(currentX, currentY) {
    let direction;
    if (prevX !== null && prevY !== null) {
      const deltaX = currentX - prevX;

      if (deltaX > sensitivity) {
        direction = 'right';
      } else if (deltaX < -sensitivity) {
        direction = 'left';
      }
    }

    /* Update the previous position to the current position */
    prevX = currentX;
    prevY = currentY;

    return direction;
  }
  
  /* Set initial draggable button position */
  function setDraggableButton() {
    
    const buttonWidth = judoDraggable.offsetWidth;
    const minScrollPos = 0;
    const maxScrollPos = judoScroller.offsetWidth - buttonWidth;
    
    let scrollPos;
    
    /* Calculate the initial scroll position based on the current image */
    const initialImageIndex = convertCurrImageNumber(options.currImage) - 1;
    const initialScrollPos = (maxScrollPos / totalImages) * initialImageIndex;
    
    scrollPos = Math.max(minScrollPos, Math.min(maxScrollPos, initialScrollPos));
    
    /* Set the initial position of the draggable button */
    judoDraggable.style.left = `${scrollPos}px`;
  }
  
  /* Handle drag move event */
  function handleDragMove(event) {
    if (dragging) {
      const judoScrollerRect = judoScroller.getBoundingClientRect();
      const buttonWidth = judoDraggable.offsetWidth;
      const minScrollPos = 0;
      const maxScrollPos = judoScroller.offsetWidth - buttonWidth;

      let scrollPos;

      if (event.type === 'touchmove') {
        scrollPos = event.touches[0].clientX - judoScrollerRect.left - buttonWidth / 2;
      } else {
        scrollPos = event.clientX - judoScrollerRect.left - buttonWidth / 2;
      }

      /* Ensure the scroll position stays within bounds */
      scrollPos = Math.max(minScrollPos, Math.min(maxScrollPos, scrollPos));

      /* Update the draggable button position */
      judoDraggable.style.left = `${scrollPos}px`;
      
      /* Calculate the current angle based on the scroll position */
      const maxRotation = 360; // 360 degrees for a full rotation
      const anglePerScroll = maxRotation / maxScrollPos;
      currentAngle = Math.floor(scrollPos * anglePerScroll);

      /* Adjust currentAngle to be within the range of 0 to 359 degrees */
      currentAngle = (currentAngle % maxRotation + maxRotation) % maxRotation;

      /* Show the corresponding image based on the current angle */
      showImageForAngle(currentAngle);
    }
  }

  /* Handle drag start event */
  function handleDragStart(event) {
    event.preventDefault();
    dragging = true;
    
    document.addEventListener('mousemove', handleDragMove);
    document.addEventListener('touchmove', handleDragMove);
  }

  /* Handle drag end event */
  function handleDragEnd() {
    dragging = false;
    /* Remove the mousemove and touchmove event listeners */
    document.removeEventListener('mousemove', handleDragMove);
    document.removeEventListener('touchmove', handleDragMove);
  }

  /* Handle draggable scroller */
  function draggableScroller() {

    /* Set judoScroller attributes and styles */
    judoScroller.id = 'judo-scroller';
    judoScroller.style.cssText = `
    position: relative;
    width: 100%;
    height: 12px;
    background-color: #f1f1f1;
    margin: 20px auto;
    border-radius: 10px;
    cursor: pointer;
    box-shadow: inset 1px 1px 3px #d7d7d7;`;

    /* Set judoDraggable attributes and styles */
    judoDraggable.id = 'judo-draggable';
    judoDraggable.style.cssText = `
    position: absolute;
    top: 0;
    left: 0;
    width: 60px;
    height: calc(100% + 10px);
    background-color: #4caf50;
    border-radius: 20px;
    cursor: pointer;
    transform: translateY(-5px);
    box-shadow: 0 0 5px rgba(0, 0, 0, 0.45);
    display: flex;
    align-items: center;
    justify-content: center;`;

    /* Create a new element (pseudo-element) */
    const pseudoElement = options.enableDragHandle ? document.createElement('span') : null;
    pseudoElement.textContent = '|||';
    pseudoElement.style.cssText = `
    color: #fff;
    position: relative;
    height: 10px;
    line-height: 10px;
    overflow: hidden;
    text-shadow: 1px 0 2px #a8a8a8;`;

    /* Common arrows CSS */
    const arrowsCommonCSS = `
    background-color: #fff;
    -webkit-mask-repeat: no-repeat;
    mask-repeat: no-repeat;
    position: absolute;
    top: 50%;
    transform: translateY(-7px);
    width: 14px;
    height: 14px;`;
    
    /* Create left arrow */
    const leftArrow = document.createElement('span');
    leftArrow.textContent = '';
    const leftIcon = `url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='14' height='14' viewBox='0 0 24 24' stroke-width='2' stroke='currentColor' fill='none' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath stroke='none' d='M0 0h24v24H0z' fill='none'%3E%3C/path%3E%3Cpath d='M13.883 5.007l.058 -.005h.118l.058 .005l.06 .009l.052 .01l.108 .032l.067 .027l.132 .07l.09 .065l.081 .073l.083 .094l.054 .077l.054 .096l.017 .036l.027 .067l.032 .108l.01 .053l.01 .06l.004 .057l.002 .059v12c0 .852 -.986 1.297 -1.623 .783l-.084 -.076l-6 -6a1 1 0 0 1 -.083 -1.32l.083 -.094l6 -6l.094 -.083l.077 -.054l.096 -.054l.036 -.017l.067 -.027l.108 -.032l.053 -.01l.06 -.01z' stroke-width='0' fill='currentColor'%3E%3C/path%3E%3C/svg%3E")`;
    leftArrow.style.cssText = `
    -webkit-mask-image: ${leftIcon};
    mask-image:${leftIcon};
    left: 1px;
    ${arrowsCommonCSS}`;
    
    /* Create right arrow */
    const rightArrow = document.createElement('span');
    rightArrow.textContent = '';
    const rightIcon = `url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='14' height='14' viewBox='0 0 24 24' stroke-width='2' stroke='currentColor' fill='none' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath stroke='none' d='M0 0h24v24H0z' fill='none'%3E%3C/path%3E%3Cpath d='M9 6c0 -.852 .986 -1.297 1.623 -.783l.084 .076l6 6a1 1 0 0 1 .083 1.32l-.083 .094l-6 6l-.094 .083l-.077 .054l-.096 .054l-.036 .017l-.067 .027l-.108 .032l-.053 .01l-.06 .01l-.057 .004l-.059 .002l-.059 -.002l-.058 -.005l-.06 -.009l-.052 -.01l-.108 -.032l-.067 -.027l-.132 -.07l-.09 -.065l-.081 -.073l-.083 -.094l-.054 -.077l-.054 -.096l-.017 -.036l-.027 -.067l-.032 -.108l-.01 -.053l-.01 -.06l-.004 -.057l-.002 -12.059z' stroke-width='0' fill='currentColor'%3E%3C/path%3E%3C/svg%3E")`;
    rightArrow.style.cssText = `
    -webkit-mask-image: ${rightIcon};
    mask-image:${rightIcon};
    right: 1px;
    ${arrowsCommonCSS}`;
    
    /* Append elements to respective parent nodes */
    judoDraggable.appendChild(pseudoElement);
    judoDraggable.appendChild(leftArrow);
    judoDraggable.appendChild(rightArrow);
    judoScroller.appendChild(judoDraggable);
    judoSpinWrapper.appendChild(judoScroller);

    /* Add event listeners to the draggable button for dragging */
    judoDraggable.addEventListener('mousedown', handleDragStart);
    judoDraggable.addEventListener('touchstart', handleDragStart);

    /* Add event listeners for mouseup and touchend events */
    document.addEventListener('mouseup', handleDragEnd);
    document.addEventListener('touchend', handleDragEnd);
  }

  if (options.enableDragHandle) {
    draggableScroller();
    setDraggableButton();
  }
}

/* Call the function for elements with the 'data-judo-spin' attribute */
const judoSpinElements = document.querySelectorAll('[data-judo-spin]');
judoSpinElements.forEach(element => judoSpin(element));